Assets/Coin.cs

```diff
@@ -2,13 +2,29 @@
 
 public class Coin : MonoBehaviour
 {
+    private CoinCounter _counter;
+    private void Awake()
+    {
+        _counter = GetComponentInParent<CoinCounter>();
```

Comment:
The approach is working.
There is an alternative: put the `CoinCounter` on the player and get it in `OnTriggerEnter`